Doc/library/functions.rst

```diff
@@ -1713,8 +1713,8 @@ are always available.  They are listed here in alphabetical order.
 
 
 .. _func-range:
-.. class:: range(stop)
-           range(start, stop, step=1)
+.. class:: range(stop, /)
```

Comment:
I have an issue with this syntax because, in this signature, `stop` is actually `start` and is only treated as `stop` because 1 arg is passed. While I understand that, in the second signature, it is intended to convey that `step` has a default value, you cannot treat the first signature the same. At least I think not. I'm happy to be proven wrong and learn something new. How would you write `start=0` in the first signature?

> If the step argument is omitted, it defaults to 1. If the start argument is omitted, it defaults to 0.

IMO, this is completely irrelevant. Yes, that is how `range` functions internally, but, for the average reader (myself included), I think it would be less confusing to just have 3 signatures if the intent is to avoid the `[, step]` syntax.

- `range(stop)`
- `range(start, stop)`
- `range(start, stop, step)`

It may also be pertinent to rephrase the description as well.

Reply:
> I have an issue with this syntax because, in this signature, stop is actually start and is only treated as stop because 1 arg is passed.

No.  Stop is stop and start is start.  The point is that this function has *multiple* signatures.  Each signature is a valid pure-python function signature.  Closest pure-python version, I think, is the typing.overload.

> How would you write start=0 in the first signature?

There is no start argument in the first signature.

> would be less confusing to just have 3 signatures if the intent is to avoid the [, step] syntax.
>    range(stop)
>    [...]

The problem is that this signature means you can do ``range(stop=123)``, which is wrong.  The ``/`` syntax might be boring, but this is a part of the language, long time ago.

Reply:
> No. Stop is stop and start is start. The point is that this function has multiple signatures. Each signature is a valid pure-python function signature. Closest pure-python version, I think, is the typing.overload.

> There is no start argument in the first signature.

Sure, but I am describing how it works internally. To me, it seems inconsistent that the description explains the behavior under-the-hood but then makes a "special case" for the second signature. Why not just have 3 and explain that in the first signature you start from 0? Why confuse the reader with "If the start argument is omitted..." as if it has a single signature?

> The problem is that this signature means you can do range(stop=123), which is wrong. The / syntax might be boring, but this is a part of the language, long time ago.

Well, you _can_, but you'll get a `TypeError`. Nothing is stopping you from doing it with `/`, either. Are we expecting folks to not understand that `stop` is a positional arg without the `/`?

Reply:
> Why confuse the reader with "If the start argument is omitted..." as if it has a single signature?

Could you propose alternative version for the documentation of range()?

> Well, you can, but you'll get a TypeError.

That means you can't.

> Nothing is stopping you from doing it with /, either.

Language syntax.  It tells me, that function doesn't support this.

Reply:
IMO, the signature(s) should be consistent with the behavior of the function and with each other.

Internally, `range` is overloaded:

- case 1: stop = args[0], start = 0, step = 1
  - `range(stop, /)`
- case 2: start = args[0], stop = args[1], step = 1
  - `range(start, stop, /)`
- case 3: start = args[0], stop = args[1], step = args[2]
  - `range(start, stop, step, /)`

I have never read a signature like `range(stop)` and thought to myself to call it like `range(stop=123)`. I have always understood, at-a-glance, that this is a positional arg and _should_ be called like `range(123)` (even though it _can_ be called like `stop=123`, just not for `range`).

The same is true of `range(start, stop, step=1)`. That reads, to me, at-a-glance, that `step` is a keyword arg, not a positional arg with a default value.

In fact, `range` agrees with me!

```py
>>> range(0, 1, step=1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: range() takes no keyword arguments
```

The only time I would read `step` as a positional arg with a default value is the signature `range(start, stop, step=1, /)`; however, you cannot represent the other signature with default values like `range(start=0, stop, step=1, /)`. This is inconsistent.

If the goal is to write signatures in such a way that language like "If the step argument is omitted, it defaults to 1. If the start argument is omitted, it defaults to 0." is not necessary then I don't think `range(stop, /)` accomplishes that. Why does one get to be "fully described" but not the other? Unfortunately, I don't think this is possible for `range` as-is.

P.S.

You linked to [4.9. More on Defining Functions](https://docs.python.org/3/tutorial/controlflow.html#more-on-defining-functions) in your [comment](https://github.com/python/cpython/issues/125897#issuecomment-2439178393) and after reading it I'm left wondering: What the hell is the difference between a positional arg with a default value and a keyword arg?

```py
>>> def foo(a=3, b=4):
...   print(a, b)
... 
>>> foo(b=2, a=1)
1 2
>>> foo(1, 2)
1 2
>>> foo()
3 4
>>> args = [5,6]
>>> foo(*args)
5 6
>>> kwargs = {"a": 7, "b": 8}
>>> foo(**kwargs)
7 8
```

According to [this](https://docs.python.org/3/glossary.html#term-argument), the answer is: "positional argument: an argument that is not a keyword argument."

😅 

I guess we should all be writing code and documentation using `/` and `*`. 

Anyway, I'm sure you are all much smarter than me and will choose the best option.

Reply:
> Internally, range is overloaded:

Yes, but your 2nd option is redundant: ``range(start, stop, step=1, /)`` covers both 2) and 3).

> I have never read a signature like range(stop) and thought to myself to call it like range(stop=123).  I have always understood, at-a-glance, that this is a positional arg and should be called like range(123)

Well, it's wrong.  Signature like ``range(stop)`` specify 1-arg functions with one mandatory positional-or-keyword argument:
```pycon
>>> def foo(spam):
...     ...
...     
>>> foo()
Traceback (most recent call last):
  File "<python-input-2>", line 1, in <module>
    foo()
    ~~~^^
TypeError: foo() missing 1 required positional argument: 'spam'
>>> foo(1)
>>> foo(spam=1)
>>> foo(bar=1)
Traceback (most recent call last):
  File "<python-input-6>", line 1, in <module>
    foo(bar=1)
    ~~~^^^^^^^
TypeError: foo() got an unexpected keyword argument 'bar'
```

> The same is true of range(start, stop, step=1). That reads, to me, at-a-glance, that step is a keyword arg, not a positional arg with a default value.

Sorry, but this is wrong as well.  All arguments in this signature are positional-or-keyword.  Last one has a default value.

> In fact, range agrees with me!

That's because it's real signature is ``range(start, stop, step=1, /)``.  Quick example:
```pycon
>>> def range(start, stop, step=1):
...     ...
...     
>>> def range2(start, stop, step=1, /):
...     ...
...     
>>> range(1, 2, step=1)
>>> range2(1, 2, step=1)
Traceback (most recent call last):
  File "<python-input-10>", line 1, in <module>
    range2(1, 2, step=1)
    ~~~~~~^^^^^^^^^^^^^^
TypeError: range2() got some positional-only arguments passed as keyword arguments: 'step'
```

> however, you cannot represent the other signature with default values like range(start=0, stop, step=1, /). This is inconsistent.

Parameters, that have default values should come last.  What's inconsistent here?

> If the goal is to write signatures 

I still don't see your alternative proposal for the range documentation.

> What the hell is the difference between a positional arg with a default value and a keyword arg?

Positional argument with a default value can be omitted, but it can't be used as a keyword argument:
```pycon
>>> def foo(bar=1, /):
...     ...
...     
>>> foo()
>>> foo(bar='spam')
Traceback (most recent call last):
  File "<python-input-17>", line 1, in <module>
    foo(bar='spam')
    ~~~^^^^^^^^^^^^
TypeError: foo() got some positional-only arguments passed as keyword arguments: 'bar'
```

> I guess we should all be writing code and documentation using / and *.

Yes, where it's required.

Reply:
Then what is your suggestion?

In the absence of `/` or `*`, should all of the references to "positional argument" and "keyword argument" in the documentation and source be changed to "required positional-or-keyword argument" and "optional positional-or-keyword argument with a default value"?

In your example:

```py
>>> def foo(spam):
...     ...
...     
>>> foo()
Traceback (most recent call last):
  File "<python-input-2>", line 1, in <module>
    foo()
    ~~~^^
TypeError: foo() missing 1 required positional argument: 'spam'
```

Here, `spam` is not a "positional argument".

Reply:
> Here, spam is not a "positional argument".

Probably, "TypeError: foo() missing 1 required argument: 'spam'" would be better here.

Reply:
@travis-pds is your argument that we shouldn't use the `/` syntax here (why?), or that we shouldn't merge the 2nd and 3rd signatures (start, stop and start, stop, step)? To me it seems that the syntax in this PR is consistent with how signatures work for functions implemented in Python, and it accurately describes the signature of `range()`.

Reply:
@JelleZijlstra I propose that we do not combine the second and third signatures.

I hope it's OK to quote from Discord. I am not holding anyone to questions answered on-the-spot. I am using this as context to continue the conversation.

Python Docs Discord:

> Carol: Documentation of `*` and `/` are not ideal... there was a decision in documentation not to redocument all existing functions as the syntax would be confusing...

> me: sorry, so i understand this correctly: it was decided to not redocument them with `*` and `/`?

> Adam: For now, I believe so. The challenge is how to best describe signatures -- we could choose the PEP 570 approach, which is consistent with runtime, but very symbol-heavy for beginners.

[From the Editorial Board decisions discussion](https://discuss.python.org/t/editorial-board-decisions/58580/3)

> ... Without the punctuation, important details about the function were missing, or had to be expressed in English...

Python Docs Discord:

> me: i could be misinterpreting but it sounded like the ultimate goal is to have the signature able to be understood without the need for a description

> Carol: Yes, in a perfect world... we made the least disruptive decision at the time of 3.8. Going forward, we would need a pragmatic solution for all users.

I see 3 options:

1. Keep the second and third signatures combined and accept that the first signature will need a description.
    a. This would require `/` in the second signature so that `step=1` is not confused as positional-or-keyword.
2. Split the second and third signatures and accept that the function will need a description.
    a. This is consistent with the C implementation and runtime behavior.
    b. The description would apply to all signatures instead of only one.
    c. This allows the flexibility to include/exclude `/`.
3. Refactor range so that there is only a need for one signature.
    a. This is the only option that wouldn't require a description but I'm not sure it's feasible.

Reply:
I don't think what you're saying should affect whether or not we can use `/`: the arguments are positional-only no matter what. It sort of sounds like from what you say that you see some connection between whether there is a default and whether a parameter is positional-only or positional-or-keyword, but that would be incorrect: in Python, all (non-variadic) parameter kind may have a default.

> Refactor range so that there is only a need for one signature

Changing the runtime behavior would be a very hard sell so I don't think it's worth discussing here.

Reply:
Maybe I should state that I am for `/`. Either way, Carol stated that it is currently undesirable. I'm not arguing for or against `/`. I am arguing that the signatures should be consistent with the goal of not needing a description and, if that can't be achieved (due to the C implementation or otherwise), then do not "hide" signatures for the sake of DRY. I think we're getting lost int he minutiae of things here.

Reply:
> I'm not arguing for or against /.

Adding `/` is all this PR does, so it seems very much on topic to talk about it. If you want a different change, this PR may not be the right place to have that conversation.

> the goal of not needing a description

I don't think we can expect the signature to be all a user needs to understand how `range()` works. The signature should show how `range()` can be called, but I don't think it's a failure if users look at the descriptive text to get a full understanding of how the class works.

Reply:
Sure.

Then I think we can resolve this with asking:
1. Is the current decision to have punctuation or not?
2. Will there be a second PR to align the functions and stdtypes pages?

Reply:
> Maybe I should state that I am for /. Either way, Carol stated that it is currently undesirable.

Could you please point on this quote?  I don't see anything like that in the referenced discussion.  While docs clearly says something different: https://devguide.python.org/documentation/style-guide/#function-signatures

Reply:
OK, so I've had a chance to review this a bit more. So, `range` is a [sequence type](https://docs.python.org/3/library/stdtypes.html#sequence-types-list-tuple-range). In this section:

> Some sequence types (such as [range](https://docs.python.org/3/library/stdtypes.html#range)) only support item sequences that follow specific patterns, and hence don’t support sequence concatenation or repetition.

In this section on types, https://docs.python.org/3/library/stdtypes.html#ranges, the docs reiterate that the `range` object is an immutable sequence type. The class constructors for `range` make sense::

- `class range(stop)`
- `class range(start, stop[, step])`

The docs clearly state the behavior of `start`, `stop` and `step`.

While the Built-in Functions docs (https://docs.python.org/3/library/functions.html#func-range) are somewhat of misnomer since `range` is still an immutable sequence type.

I think if I were to make any change for clarity I would change the Built-in Functions page to match the types page.

- `class range(stop)`
- `class range(start, stop[, step])`

The `/` and `*` don't apply to `range` as it is an immutable sequence type and not a true function.

If you would like to amend the Built-in Functions page for `range` to remove `step=1` and add the wording about the defaults like in the types doc, that would be acceptable to me.

Reply:
> I would change the Built-in Functions page to match the types page.

That should work too, as the source of confusion was ``range(start, stop, step=1)`` signature, which suggest that ``step`` is optional positional-or-keyword argument.

On another hand, syntax like ``range(start, stop[, step])`` is not a valid Python syntax.  It's even not explained somewhere in docs.  So, such signatures also might add problems, especially for beginners.

> The / and * don't apply to range as it is an immutable sequence type and not a true function.

For classes, signature means the one for ``__new__`` or ``__init__`` dunder method.  E.g.:
```pycon
>>> class A:
...     def __init__(self, val):
...         self._val = val
... 
>>> help(A)
Help on class A in module __main__:

class A(builtins.object)
 |  A(val)
 |
 |  Methods defined here:
 |
 |  __init__(self, val)
 |      Initialize self.  See help(type(self)) for accurate signature.
 |
 |  ----------------------------------------------------------------------
 |  Data descriptors defined here:
 |
 |  __dict__
 |      dictionary for instance variables
 |
 |  __weakref__
 |      list of weak references to the object

>>> import inspect
>>> inspect.signature(A)
<Signature (val)>
```

So, this funny syntax is applicable for types as well.

The only problem is that range() has multiple signatures.  This is something that the inspect module can't handle (see WIP in https://github.com/python/cpython/pull/117671).

Reply:
Excellent point @skirpichev about `range(start, stop[, step])`.  What do you recommend that we do in the short term before Serhiy's PR is complete and merged?

P.S. I liked your thoughtful response and detail. Thanks!

Reply:
I think that for multiple signatures - every possible variant should be represented by valid Python syntax, like in this pr.

Using ``range(start, stop[, step])``-like syntax seems to be a dead end.  While this notation seems obvious e.g. to me, I'm not sure it is simple for newbies.

Where we can document it to be discoverable?  I don't see a good place.  Perhaps, here: https://docs.python.org/3/reference/introduction.html#notation.  But this is used in the stdlib as well.

Reply:
> > Maybe I should state that I am for /. Either way, Carol stated that it is currently undesirable.
> 
> Could you please point on this quote? I don't see anything like that in the referenced discussion. While docs clearly says something different: https://devguide.python.org/documentation/style-guide/#function-signatures

@skirpichev  It was in a conversation in the Python Docs Discord.

> "Documentation of `*` and `/` in functions are not ideal. When positional only parameters were introduced in 3.8 and PEP 570 [link], there was a decision in documentation not to redocument all existing functions as the syntax would be confusing. It's not ideal and it's something that we can revisit, but it would be on a documentation wide decision not an individual PR decision."

Reply:
To me it still seems inconsistent to only consider range here, since https://github.com/python/cpython/issues/91485 / PR https://github.com/python/cpython/pull/96579, dropped the `[]` notation in favor of `/` which were then dropped in https://github.com/python/cpython/issues/98340 / PR https://github.com/python/cpython/pull/99476 **for a bunch of functions**.

This seems to be a much broader documentation topic to me.

Reply:
It was decided in python/devguide#1344 to include the slashes, superseding those old PRs.

Reply:
Thanks @JelleZijlstra. @travis-pds Hmm... I tend to take the approach of user first. Undesirable is too strong a word. It was not ideal when 3.8 was released. Time has passed and @JelleZijlstra has captured well the current thinking. I tend to read most of the decisions as in most cases... yet, it may make sense in some cases to do something else if it benefits readability and comprehension.

Reply:
> It was decided in [python/devguide#1344](https://github.com/python/devguide/pull/1344) to include the slashes, superseding those old PRs.

Hence I am asking why only `range` should get the slashes again and not all other functions that lost them in https://github.com/python/cpython/pull/99476?

Reply:
> To me it still seems inconsistent to only consider range here

@chris-eibl, this pr address concrete issue with range, not about mechanical reversion of https://github.com/python/cpython/pull/99476.

> Like e.g. bool(x=False, /)?

Please open issue.

Reply:
Just for `bool`? This does not seem consistent to me, but the mechanical reversion does not seem to be liked either.
Therefore, I leave it to you core dev(s) to decide which and how many issues to create, since there are more canditates than `bool`.